astropy/utils/misc.py

```diff
@@ -382,18 +384,43 @@ class JsonCustomEncoder(json.JSONEncoder):
 
     """
     def default(self, obj):
+        from .. import units as u
         import numpy as np
-        if isinstance(obj, (np.ndarray, np.number)):
+        if isinstance(obj, np.number):
+            return obj.tolist()
+        elif isinstance(obj, u.Quantity):
+            return dict(value=obj.value, unit=obj.unit.to_string())
+        elif isinstance(obj, np.ndarray):  # Must come after Quantity
             return obj.tolist()
         elif isinstance(obj, (complex, np.complex)):
             return [obj.real, obj.imag]
         elif isinstance(obj, set):
             return list(obj)
         elif isinstance(obj, bytes):  # pragma: py3
             return obj.decode()
+        elif isinstance(obj, u.UnitBase):
+            return obj.name
+        elif obj is u.dimensionless_unscaled:
```

Comment:
Since `isinstance(u.dimensionless_unscaled, u.UnitBase)` is true, this needs to be rewritten (and tested!). Since it also is not a unique unit, I would write:
```
elif isinstance(obj, (u.UnitBase, u.FunctionUnitBase)):
    if obj == u.dimensionless_unscaled:
        obj = 'dimensionless_unit'
    else:
        obj = obj.to_string()
```
(note the inclusion of function units here -- we do need #3650... -- as well as the use of `to_string` instead of `name`, which I think is more reliable).